### DISCUSSION TITLE
Difference Between LOGIN Event Type and GRANT_CONSENT

### DISCUSSION THREAD

**Discussion by brunoguic** [keycloak/keycloak#32771] [Q&A] [answered]
Hello,

I have an OpenID Connect client, and I want to measure the number of users who have granted consent to this client.

I expected to see the event type GRANT_CONSENT, but I only see LOGIN and CODE_TO_TOKEN. After reviewing the code, I noticed that the GRANT_CONSENT event is only triggered by the Account REST API (refer to this  #11199)

My questions are:

- Is the LOGIN event type what I should expect for measuring this metric?
- Should the GRANT_CONSENT event be triggered along with the LOGIN event?
	
Thanks

**Comment by sschu** [comment, marked as answer]
Currently, you need to check for LOGIN events that contain a granted consent plus GRANT_CONSENT events that came from the Account Rest API. Always creating a GRANT_CONSENT event would make searching for consents more straight-forward. However, AFAIK only one event is normally generated at a time so changing this would deviate from this principle.

**Comment by brunoguic** [comment]
Thanks @sschu

**Comment by erickmoreno** [comment]
I understand your answer @sschu. But, in this case, should we enrich the LOGIN event to have more GRANT_CONSENT details?

I think that the single event principle can be important, but we may need more data to make those single events really useful.  
See this example. LOGIN event only says that some consent was granted, but it didn't say which one.

If we need to track consents granted, but not only the number, the consent itself, we need to filter all logins and another event like CODE_TO_TOKEN. 

This makes things harder, since the relationship between both events may not be easy to define on a busy environment.

**Reply by sschu** [reply]
Sure, the current situation is not ideal. I think it makes sense to have the same information available in both events. However, how is CODE_TO_TOKEN related to this? AFAIK it doesnt contain any consent-related data.

**Reply by erickmoreno** [reply]
`CODE_TO_TOKEN` event has the `SCOPE` field that can be used to see which consents were given.

**Reply by sschu** [reply]
Ah, makes sense. I realized there are other places where more than one event is emitted per request so it might be feasible to also emit GRANT_CONSENT on login.
